modules/calib3d/src/solvepnp.cpp

```diff
@@ -456,4 +398,325 @@ int solveP3P( InputArray _opoints, InputArray _ipoints,
+    {
+        CV_Assert(npoints == 4);
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+        p3p P3Psolver(cameraMatrix);
+
+        Mat rvec, tvec, R;
+        bool result = P3Psolver.solve(R, tvec, opoints, undistortedPoints);
+        if (result)
+        {
+            Rodrigues(R, rvec);
+            vec_rvecs.push_back(rvec);
+            vec_tvecs.push_back(tvec);
+        }
+    }
+    else if (flags == SOLVEPNP_AP3P)
+    {
+        CV_Assert(npoints == 4);
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+        ap3p P3Psolver(cameraMatrix);
+
+        Mat rvec, tvec, R;
+        bool result = P3Psolver.solve(R, tvec, opoints, undistortedPoints);
+        if (result)
+        {
+            Rodrigues(R, rvec);
+            vec_rvecs.push_back(rvec);
+            vec_tvecs.push_back(tvec);
+        }
+    }
+    else if (flags == SOLVEPNP_ITERATIVE)
+    {
+        Mat rvec, tvec;
+        if (useExtrinsicGuess)
+        {
+            rvec = _rvec.getMat();
+            tvec = _tvec.getMat();
+        }
+        else
+        {
+            rvec.create(3, 1, mtype);
+            tvec.create(3, 1, mtype);
+        }
+
+        CvMat c_objectPoints = cvMat(opoints), c_imagePoints = cvMat(ipoints);
+        CvMat c_cameraMatrix = cvMat(cameraMatrix), c_distCoeffs = cvMat(distCoeffs);
+        CvMat c_rvec = cvMat(rvec), c_tvec = cvMat(tvec);
+        cvFindExtrinsicCameraParams2(&c_objectPoints, &c_imagePoints, &c_cameraMatrix,
+                                     (c_distCoeffs.rows && c_distCoeffs.cols) ? &c_distCoeffs : 0,
+                                     &c_rvec, &c_tvec, useExtrinsicGuess );
+
+        vec_rvecs.push_back(rvec);
+        vec_tvecs.push_back(tvec);
+    }
+    else if (flags == SOLVEPNP_IPPE)
+    {
+        double val = 0.0;
+        bool planar = isPlanarObjectPoints(opoints, 1e-5, val);
+        if (planar)
+        {
+            Mat undistortedPoints;
+            undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+
+            IPPE::PoseSolver poseSolver;
+            Mat rvec1, tvec1, rvec2, tvec2;
+            float reprojErr1, reprojErr2;
+            poseSolver.solveGeneric(opoints, undistortedPoints, rvec1, tvec1, reprojErr1, rvec2, tvec2, reprojErr2);
+
+            if (reprojErr1 < reprojErr2)
+            {
+                vec_rvecs.push_back(rvec1);
+                vec_tvecs.push_back(tvec1);
+
+                vec_rvecs.push_back(rvec2);
+                vec_tvecs.push_back(tvec2);
+            }
+            else
+            {
+                vec_rvecs.push_back(rvec2);
+                vec_tvecs.push_back(tvec2);
+
+                vec_rvecs.push_back(rvec1);
+                vec_tvecs.push_back(tvec1);
+            }
+        }
+    }
+    else if (flags == SOLVEPNP_IPPE_SQUARE)
+    {
+        CV_Assert(npoints == 4);
```

Comment:
`SOLVEPNP_IPPE_SQUARE` requires the object point to be defined in a specific manner:
```
[-squareLength / 2.0, squareLength / 2.0, 0]
[squareLength / 2.0, squareLength / 2.0, 0]
[squareLength / 2.0, -squareLength / 2.0, 0]
[-squareLength / 2.0, -squareLength / 2.0, 0]
```

I have added some checks. Ultimately, it will be slower to call `solvePnP` with `SOLVEPNP_IPPE_SQUARE` than in the original method.

Reply:
Agreed :+1: